src/derive.rs

```diff
@@ -273,25 +273,31 @@ pub trait Subcommand: FromArgMatches + Sized {
 ///     level: Level,
 /// }
 ///
-/// #[derive(clap::ArgEnum)]
+/// #[derive(clap::ArgEnum, Clone)]
 /// enum Level {
 ///     Debug,
 ///     Info,
 ///     Warning,
 ///     Error,
 /// }
 /// ```
-pub trait ArgEnum: Sized {
-    /// All possible argument choices, in display order.
-    const VARIANTS: &'static [&'static str];
+pub trait ArgEnum: Sized + Clone {
+    /// All possible argument values, in display order.
+    fn value_variants<'a>() -> &'a [Self];
```

Comment:
@epage With the recent const improvements, do you think we change the return to be `[Self; T]` where T is the number of variants?

Reply:
I don't think so, const generics are very limited.

I gave it a try on the playground and I didn't see a way to make it work.
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=75afb67dc29136f6341e6ea4f044fea9

Reply:
Could actually be possible, I tried that to replace the variants slice befor, the issue I had, was I didn't find all the places where I needed to specify the generic. But I can try again, maybe you get more out of the error messages then me.

Reply:
This worked though. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8ccbfd753fb15dab675ba0491c7f48a3

Reply:
Yes, but that severally limits how the trait can work because all calls to it have to be generic over the lenth.

Reply:
I don't see many calls at all. Am I missing something?

Reply:
this would not solve the `Clone` requirenment, well maybe in 2021 but not yet if I understand this correctly:
```
this method call resolves to `<&[T; N] as IntoIterator>::into_iter` (due to backwards
 compatibility), but will resolve to <[T; N] as IntoIterator>::into_iter in Rust 2021.
    `#[warn(array_into_iter)]` on by default
    this changes meaning in Rust 2021
    for more information, see issue #66145 <https://github.com/rust-lang/rust/issues/66145>
    ```

Reply:
@ModProg, depending on what the code looks like, you can force it to use the correct `IntoIterator` without the new edition

@pksunkara Yes, there aren't many calls but we have to weigh out the API design for which is less likely to break us.  A `Clone` is not a heavy requirement to put on these types.  A per implementation const generic severely limits how the trait can be used.

However, as I said before, I think resolving of this needs to be split out into its own issue and that we should merge this as-is.  As we've seen through these discussions, there is a lot of nuance to the implementation for any proposed solution, so rather than continue having a lot of back-and-forth in these PRs, we should merge this as-is rather than block it on incremental improvements and allow someone to focus on both the API design and implementation to reduce back and forths on this.